syntax = "proto3";
import "google/protobuf/wrappers.proto";

package core.protos;

message TimeRange {
  uint64 time_geq = 1;
  uint64 time_lt = 2;
}

message Page {
  google.protobuf.UInt32Value limit = 1;
  google.protobuf.UInt32Value offset = 2;
}

enum Function {
  UNKNOWN_F = 0;
  AVG_F = 1;
  SUM_F = 2;
  MIN_F = 3;
  MAX_F = 4;
  LAST_F = 5;
}

enum Operator {
  UNKNOWN_O = 0;
  GREATER_THAN_O = 1;
  GREATER_THAN_EQUAL_O = 2;
  LESS_THAN_O = 3;
  LESS_THAN_EQUAL_O = 4;
  EQUAL_O = 5;
  NOT_EQUAL_O = 6;
  LIKE_O = 7;
  EXISTS_O = 8;
  CONTAINS_O = 9;
}

enum LogicalOperator {
  UNKNOWN_LO = 0;
  AND_LO = 1;
  OR_LO = 2;
  NOT_LO = 3;
}

enum Op {
  UNKNOWN_OP = 0;
  EQ = 1;
  NEQ = 2;
  GT = 3;
  LT = 4;
  GTE = 5;
  LTE = 6;
  IN = 7;
  NOT_IN = 8;
  IS_NULL = 9;
  EXISTS = 10;

  AND = 20;
  OR = 21;
  NOT = 22;
}

message OpDescription {
  Op op = 1;
  string label = 2;
  bool is_unary = 3;
  bool is_logical = 4;
}

message TaskCronSchedule {
  google.protobuf.StringValue minutes = 1;
  google.protobuf.StringValue hours = 2;
  google.protobuf.StringValue days_of_the_week = 3;
  google.protobuf.StringValue days_of_the_month = 4;
  google.protobuf.StringValue days_of_the_year = 5;
  google.protobuf.StringValue timezone = 6;
}

message TaskInterval {
  google.protobuf.UInt64Value interval_in_seconds = 1;
}

message TaskCronRule {
  google.protobuf.StringValue rule = 1;
  google.protobuf.StringValue timezone = 2;
}

enum Source {
  UNKNOWN = 0;
  SENTRY = 1;
  SEGMENT = 2;
  ELASTIC_SEARCH = 3;
  AMPLITUDE = 4;
  AWS_KINESIS = 5;
  CLOUDWATCH = 6;
  CLEVERTAP = 7;
  RUDDERSTACK = 8;
  MOENGAGE = 9;
  CRIBL = 10;
  KAFKA = 11;
  DATADOG = 12;
  FILEBEAT = 13;
  LOGSTASH = 14;
  FLUENTD = 15;
  FLUENTBIT = 16;
  PAGER_DUTY = 17;
  NEW_RELIC = 18;
  SLACK = 19;
  HONEYBADGER = 20;
  GOOGLE_CHAT = 21;
  DATADOG_OAUTH = 22;
  GCM = 23;
  PROMETHEUS = 24;
  ELASTIC_APM = 25;
  VICTORIA_METRICS = 26;
  SLACK_CONNECT = 27;
  GRAFANA = 28;
  CLICKHOUSE = 29;
  DOCUMENTATION = 30;
  POSTGRES = 31;
  OPS_GENIE = 32;
  EKS = 33;
  AGENT_PROXY = 34;
  GRAFANA_VPC = 35;
  GITHUB = 36;
  SQL_DATABASE_CONNECTION = 37;
  OPEN_AI = 38;
  REMOTE_SERVER = 39;
  API = 40;
  BASH = 41;
  AZURE = 42;
  GRAFANA_MIMIR = 43;
  GKE = 44;
  MS_TEAMS = 45;
  GRAFANA_LOKI = 46;
  KUBERNETES = 47;
  SMTP = 48;
  BIG_QUERY = 49;
  ZENDUTY = 50;
  ROOTLY = 51;
  JIRA_CLOUD = 52;
  ASANA = 53;
  CONFLUENCE_CLOUD = 54;
  CONFLUENCE_SELF_HOSTED = 55;
  GOOGLE_DRIVE = 56;
  NOTION = 57;
  DRD_PROXY_AGENT = 58;
  MONGODB = 59;
  OPEN_SEARCH = 60;
  LINEAR = 61;
  JENKINS = 62;
  GITHUB_ACTIONS = 63;
  CUSTOM_STRATEGIES = 64;
  ARGOCD = 65;
  DRD_ALERT_WEBHOOK = 66;
  ROLLBAR = 67;
  LAMBDA = 68;
  BROWSER = 69;
  POSTHOG = 70;
  MIXPANEL = 71;
  SIGNOZ = 72;
  BITBUCKET = 73;
  CORALOGIX = 74;
  RENDER = 75;
  MCP_SERVER = 76;
  VICTORIA_LOGS=77;
}

enum SourceKeyType {
  UNKNOWN_SKT = 0;
  SENTRY_API_KEY = 1;
  SENTRY_ORG_SLUG = 6;

  DATADOG_APP_KEY = 2;
  DATADOG_API_KEY = 3;
  DATADOG_AUTH_TOKEN = 15;
  DATADOG_API_DOMAIN = 18;

  NEWRELIC_API_KEY = 4;
  NEWRELIC_APP_ID = 5;
  NEWRELIC_QUERY_KEY = 7;
  NEWRELIC_API_DOMAIN = 19;

  SLACK_BOT_AUTH_TOKEN = 8;
  SLACK_CHANNEL_SKT = 9;
  SLACK_APP_ID = 46;
  SLACK_APP_CLIENT_ID = 141;
  SLACK_APP_CLIENT_SECRET = 142;
  SLACK_APP_SIGNING_SECRET = 143;

  HONEYBADGER_USERNAME = 10;
  HONEYBADGER_PASSWORD = 11;
  HONEYBADGER_PROJECT_ID = 12;

  AWS_ACCESS_KEY = 13;
  AWS_SECRET_KEY = 14;
  AWS_REGION = 20;
  AWS_ASSUMED_ROLE_ARN = 23;
  EKS_ROLE_ARN = 40;
  AWS_DRD_CLOUD_ROLE_ARN = 93;

  GOOGLE_CHAT_BOT_OAUTH_TOKEN = 16;
  GOOGLE_CHAT_BOT_SPACES = 17;

  GRAFANA_HOST = 21;
  GRAFANA_API_KEY = 22;
  GRAFANA_TEAM_HOST = 153;

  CLICKHOUSE_INTERFACE = 24;
  CLICKHOUSE_HOST = 25;
  CLICKHOUSE_PORT = 26;
  CLICKHOUSE_USER = 27;
  CLICKHOUSE_PASSWORD = 28;

  GCM_PROJECT_ID = 29;
  GCM_PRIVATE_KEY = 30;
  GCM_CLIENT_EMAIL = 31;
  GCM_TOKEN_URI = 32;
  GCM_SERVICE_ACCOUNT_JSON = 112;

  POSTGRES_HOST = 33;
  POSTGRES_USER = 34;
  POSTGRES_PASSWORD = 35;
  POSTGRES_PORT = 36;
  POSTGRES_DATABASE = 37;
  POSTGRES_OPTIONS = 38;

  SQL_DATABASE_CONNECTION_STRING_URI = 39;

  PAGER_DUTY_API_KEY = 41;
  PAGER_DUTY_CONFIGURED_EMAIL = 63;

  OPS_GENIE_API_KEY = 42;

  AGENT_PROXY_HOST = 43;
  AGENT_PROXY_API_KEY = 44;

  GITHUB_TOKEN = 45;
  GITHUB_ORG = 105;

  OPEN_AI_API_KEY = 47;

  REMOTE_SERVER_PEM = 49;
  REMOTE_SERVER_USER = 50;
  REMOTE_SERVER_HOST = 51;
  REMOTE_SERVER_PASSWORD = 52;
  REMOTE_SERVER_PORT = 111;

  MIMIR_HOST = 53;
  X_SCOPE_ORG_ID = 54;

  SSL_VERIFY = 55;

  AZURE_SUBSCRIPTION_ID = 56;
  AZURE_TENANT_ID = 57;
  AZURE_CLIENT_ID = 58;
  AZURE_CLIENT_SECRET = 59;

  GKE_PROJECT_ID = 60;
  GKE_SERVICE_ACCOUNT_JSON = 61;

  MS_TEAMS_CONNECTOR_WEBHOOK_URL = 62;

  ELASTIC_SEARCH_PROTOCOL = 64;
  ELASTIC_SEARCH_HOST = 65;
  ELASTIC_SEARCH_PORT = 66;
  ELASTIC_SEARCH_API_KEY_ID = 67;
  ELASTIC_SEARCH_API_KEY = 68;

  GRAFANA_LOKI_PROTOCOL = 69;
  GRAFANA_LOKI_HOST = 70;
  GRAFANA_LOKI_PORT = 71;

  KUBERNETES_CLUSTER_NAME = 72;
  KUBERNETES_CLUSTER_API_SERVER = 73;
  KUBERNETES_CLUSTER_TOKEN = 74;
  KUBERNETES_CLUSTER_CERTIFICATE_AUTHORITY_DATA = 75;
  KUBERNETES_CLUSTER_CERTIFICATE_AUTHORITY_PATH = 76;

  SMTP_HOST = 77;
  SMTP_PORT = 78;
  SMTP_USER = 79;
  SMTP_PASSWORD = 80;

  BIG_QUERY_PROJECT_ID = 81;
  BIG_QUERY_SERVICE_ACCOUNT_JSON = 82;

  ZENDUTY_API_KEY = 83;

  ROOTLY_API_KEY = 84;

  JIRA_CLOUD_API_KEY = 85;
  JIRA_DOMAIN = 109;
  JIRA_EMAIL = 110;

  ASANA_ACCESS_TOKEN = 86;

  CONFLUENCE_CLOUD_API_KEY = 87;
  CONFLUENCE_CLOUD_SITE_URL = 139;
  CONFLUENCE_CLOUD_ATLASSIAN_ACCOUNT_EMAIL = 140;

  NOTION_API_KEY = 88;

  DRD_PROXY_AGENT_HOST = 89;
  DRD_PROXY_AGENT_API_TOKEN = 90;
  DRD_PROXY_AGENT_API_KEY = 91;

  MONGODB_CONNECTION_STRING = 92;

  OPEN_SEARCH_PROTOCOL = 94;
  OPEN_SEARCH_HOST = 95;
  OPEN_SEARCH_PORT = 96;
  OPEN_SEARCH_USERNAME = 97;
  OPEN_SEARCH_PASSWORD = 98;

  LINEAR_API_KEY = 99;

  JENKINS_URL = 100;
  JENKINS_USERNAME = 101;
  JENKINS_API_TOKEN = 102;
  JENKINS_CRUMB = 113;

  GITHUB_ACTIONS_TOKEN = 103;

  CUSTOM_STRATEGIES_ACCOUNT_ID = 104;

  ARGOCD_SERVER = 106;
  ARGOCD_TOKEN = 107;

  ROLLBAR_ACCESS_TOKEN = 108;

  POSTHOG_API_KEY = 121;
  POSTHOG_APP_HOST = 122;
  POSTHOG_PROJECT_ID = 123;

  MIXPANEL_SERVICE_ACCOUNT_USERNAME = 131;
  MIXPANEL_SERVICE_ACCOUNT_SECRET = 132;
  MIXPANEL_DOMAIN = 133;
  MIXPANEL_PROJECT_ID = 134;
  MIXPANEL_WORKSPACE_ID = 135;

  SIGNOZ_API_URL = 136;
  SIGNOZ_API_TOKEN = 137;

  KIBANA_HOST = 138;

  // 139, 140 is used
  BITBUCKET_WORKSPACE = 144;
  BITBUCKET_API_KEY = 145;
  BITBUCKET_REPO = 146;
  RENDER_API_KEY = 147;
  
  CORALOGIX_API_KEY = 148;
  CORALOGIX_ENDPOINT = 149;
  CORALOGIX_DOMAIN = 152;
  
  MCP_SERVER_BASE_URL = 150;
  MCP_SERVER_AUTH_HEADERS = 151;

  // VictoriaLogs specific keys (used by VictoriaLogs connector forms)
  VICTORIA_LOGS_PROTOCOL = 154;
  VICTORIA_LOGS_HOST = 155;
  VICTORIA_LOGS_PORT = 156;
  VICTORIA_LOGS_HEADERS = 157;

}

enum SourceModelType {
  UNKNOWN_MT = 0;
  //New Relic Models
  NEW_RELIC_POLICY = 1;
  NEW_RELIC_CONDITION = 2;
  NEW_RELIC_ENTITY = 3;
  NEW_RELIC_ENTITY_DASHBOARD = 4;
  NEW_RELIC_ENTITY_APPLICATION = 5;
  NEW_RELIC_NRQL = 6;
  NEW_RELIC_ENTITY_DASHBOARD_V2 = 7;

  //Datadog Models
  DATADOG_MONITOR = 101;
  DATADOG_DASHBOARD = 102;
  DATADOG_LIVE_INTEGRATION_AWS = 103;
  DATADOG_LIVE_INTEGRATION_AWS_LOG = 104;
  DATADOG_LIVE_INTEGRATION_AZURE = 105;
  DATADOG_LIVE_INTEGRATION_CLOUDFLARE = 106;
  DATADOG_LIVE_INTEGRATION_FASTLY = 107;
  DATADOG_LIVE_INTEGRATION_GCP = 108;
  DATADOG_LIVE_INTEGRATION_CONFLUENT = 109;
  DATADOG_SERVICE = 110;
  DATADOG_METRIC = 111;
  DATADOG_QUERY = 112;
  DATADOG_APM = 113;
  DATADOG_FIELDS = 114;

  //Cloudwatch Models
  CLOUDWATCH_METRIC = 201;
  CLOUDWATCH_LOG_GROUP = 202;
  CLOUDWATCH_LOG_GROUP_QUERY = 203;
  CLOUDWATCH_ALARMS = 204;
  RDS_INSTANCES = 205;
  CLOUDWATCH_DASHBOARD = 206;

  //Grafana Models
  GRAFANA_DATASOURCE = 301;
  GRAFANA_DASHBOARD = 302;
  GRAFANA_TARGET_METRIC_PROMQL = 303;
  GRAFANA_PROMETHEUS_DATASOURCE = 304;
  GRAFANA_ALERT_RULE = 305;
  GRAFANA_LOKI_DATASOURCE = 306;

  //Clickhouse Models
  CLICKHOUSE_DATABASE = 401;
  CLICKHOUSE_TABLE = 402;

  //Slack Models
  SLACK_CHANNEL = 501;

  //Documentation Models
  MARKDOWN = 601;
  IFRAME = 602;

  //Postgres Models
  POSTGRES_QUERY = 701;
  POSTGRES_TABLE = 702;

  //EKS Models
  EKS_CLUSTER = 801;
  EKS_NAMESPACE = 802;
  EKS_SERVICE = 803;
  EKS_DEPLOYMENT = 804;
  EKS_INGRESS = 805;
  EKS_NETWORK_POLICY = 806;
  EKS_HPA = 807;
  EKS_REPLICASET = 808;
  EKS_STATEFULSET = 809;

  //Sql Database Connection Models
  SQL_DATABASE_CONNECTION_RAW_QUERY = 901;
  SQL_DATABASE_TABLE = 902;

  //Azure Models
  AZURE_WORKSPACE = 1001;
  AZURE_RESOURCE = 1002;

  //Remote Server Models
  SSH_SERVER = 1100;
  SSH_USER = 1101;

  //Mimir Server Models
  GRAFANA_MIMIR_PROMQL = 1201;

  //GKE Models
  GKE_CLUSTER = 1301;
  GKE_NAMESPACE = 1302;
  GKE_SERVICE = 1303;
  GKE_DEPLOYMENT = 1304;
  GKE_INGRESS = 1305;
  GKE_NETWORK_POLICY = 1306;
  GKE_HPA = 1307;
  GKE_REPLICASET = 1308;
  GKE_STATEFULSET = 1309;

  //MS TEAMS Models
  MS_TEAMS_CHANNEL = 1401;

  //Pagerduty Models
  PAGERDUTY_INCIDENT = 1501;

  //Elastic Search Models
  ELASTIC_SEARCH_INDEX = 1601;
  ELASTIC_SEARCH_SERVICES = 1602;
  ELASTIC_SEARCH_DASHBOARDS = 1603;
  ELASTIC_SEARCH_INDEX_PATTERNS = 1604;

  //GCM Models
  GCM_METRIC = 1701;
  GCM_DASHBOARD = 1702;
  GCM_CLOUD_RUN_SERVICE_DASHBOARD = 1703;

  //Zen Duty Models
  ZENDUTY_INCIDENT = 1801;

  //ROOTLY Models
  ROOTLY_INCIDENT = 1901;

  //OPS GENIE Models
  OPS_GENIE_ESCALATION = 2001;
  OPS_GENIE_TEAM = 2002;

  //Google Chat Models
  GOOGLE_CHAT_SPACE = 2101;

  // MongoDB Models
  MONGODB_DATABASE = 2201;
  MONGODB_COLLECTION = 2202;

  //Open Search Models
  OPEN_SEARCH_INDEX = 2301;

  // Asana Models
  ASANA_PROJECT = 2401;

  // Custom Strategies
  MODELS_ACTIVE = 2501;

  //Github Models
  GITHUB_REPOSITORY = 2601;
  GITHUB_MEMBER = 2602;

  // Jira Models
  JIRA_PROJECT = 2701;
  JIRA_USER = 2702;

  // ArgoCd Models
  ARGOCD_APPS = 2801;

  // jenkins Models
  JENKINS_JOBS = 2901;

  // PostHog Models
  POSTHOG_PERSON = 3001;
  POSTHOG_GROUP = 3002;
  POSTHOG_COHORT = 3003;
  POSTHOG_PROPERTY = 3004;

  // ECS Models
  ECS_CLUSTER = 4001;
  ECS_TASK = 4002;
  ECS_SERVICE = 4003;
  
  // SignOz Models
  SIGNOZ_DASHBOARD = 5001;
  SIGNOZ_ALERT = 5002;
  SIGNOZ_METRIC = 5003;
  SIGNOZ_SERVICE = 5004;
  SIGNOZ_LOG_ATTRIBUTES = 5005;

  // Kubernetes Models
  KUBERNETES_NAMESPACE = 6001;
  KUBERNETES_SERVICE = 6002;
  KUBERNETES_DEPLOYMENT = 6003;
  KUBERNETES_INGRESS = 6004;
  KUBERNETES_NETWORK_POLICY = 6005;
  KUBERNETES_HPA = 6006;
  KUBERNETES_REPLICASET = 6007;
  KUBERNETES_STATEFULSET = 6008;
  KUBERNETES_NETWORK_MAP = 6009;

  SENTRY_PROJECT = 7001;

  // Coralogix Models
  CORALOGIX_DASHBOARD = 8001;
  CORALOGIX_INDEX_MAPPING = 8002;

  // VictoriaLogs Models
  VICTORIA_LOGS_FIELD = 9001;
}

message EventKey {
  enum KeyType {
    UNKNOWN = 0;
    STRING = 1;
    LONG = 2;
    DOUBLE = 3;
    BOOLEAN = 4;
    BYTE = 5;
    ARRAY = 6;
    OBJECT = 7;
  }
  uint64 id = 1;
  string key = 2;
  KeyType key_type = 3;
  EventTypePartial event_type = 4;
}

message EventType {
  uint64 id = 1;
  string name = 2;
  repeated EventKey keys = 3;
  repeated Event.EventSource event_sources = 4;
}

message EventTypeStats {
  google.protobuf.UInt64Value keys_count = 1;
  google.protobuf.UInt64Value event_count = 2;
  google.protobuf.UInt64Value monitor_count = 3;
}

message EventTypePartial {
  uint64 id = 1;
  string name = 2;
}

message EventTypeSummary {
  EventTypePartial event_type = 1;
  EventTypeStats stats = 2;
}

message EventTypeDefinition {
  EventType event_type = 1;
  EventTypeStats stats = 2;
}

message Event {
  enum EventSource {
    UNKNOWN = 0;
    SAMPLE = 1;
    API = 2;
    SDK = 3;
    SEGMENT = 4;
    AMPLITUDE = 5;
    SNS = 6;
    CLOUDWATCH = 7;
    COLLECTOR = 8;
    AWS_KINESIS = 9;
    SLACK = 10;
    OPS_GENIE = 11;
  }
  uint64 id = 1;
  EventTypePartial event_type = 2;
  repeated KeyValue kvs = 3;
  sfixed64 timestamp = 4;
  EventSource event_source = 5;
}

enum Context {
  UNKNOWN_C = 0;
  EVENT = 1;
  MONITOR_TRANSACTION = 2;
  ENTITY_INSTANCE = 3;
  EVENT_TYPE = 4;
  MONITOR = 5;
  ENTITY = 6;
  EVENTS_CLICKHOUSE = 7;
  PANEL_NAME = 8;
  DASHBOARD_NAME = 9;
  EVENTS_ALERT_OPS_CLICKHOUSE = 10;
  SLACK_ALERT = 11;
  EVENTS_PRODIGAL_CLICKHOUSE = 12;
  ALERT_OPS = 13;
  ALERT_MIDDLEWARE = 14;
}

message Value {
  // The value is one of the listed fields. It is valid for all values to be unspecified
  // in which case this AnyValue is considered to be "empty".
  oneof value {
    string string_value = 1;
    bool bool_value = 2;
    int64 int_value = 3;
    double double_value = 4;
    ArrayValue array_value = 5;
    KeyValueList kvlist_value = 6;
    bytes bytes_value = 7;
  }
}

message ArrayValue {
  // Array of values. The array may be empty (contain 0 elements).
  repeated Value values = 1;
}

message KeyValueList {
  // A collection of key/value pairs. The list may be empty (may
  // contain 0 elements).
  // The keys MUST be unique (it is not allowed to have more than one
  // value with the same key).
  repeated KeyValue values = 1;
}

message KeyValue {
  string key = 1;
  Value value = 2;
}

// Recent Artifact Types for recent_changes tracking
// These are used to categorize different types of artifacts tracked in the recent_changes table
enum RecentArtifactType {
  RA_UNKNOWN = 0;

  // GitHub Artifacts
  RA_GITHUB_REPO = 1;
  RA_GITHUB_COMMIT = 2;
  RA_GITHUB_PR = 3;
  RA_GITHUB_RELEASE = 4;

  // GitLab Artifacts (future)
  RA_GITLAB_REPO = 101;
  RA_GITLAB_COMMIT = 102;
  RA_GITLAB_MR = 103;
  RA_GITLAB_RELEASE = 104;

  // Bitbucket Artifacts (future)
  RA_BITBUCKET_REPO = 201;
  RA_BITBUCKET_COMMIT = 202;
  RA_BITBUCKET_PR = 203;
}